src/main/java/javax/measure/unit/Unit.java

```diff
@@ -211,6 +248,34 @@ public final Dimension getDimension() {
      */
     public final UnitConverter getConverterTo(Unit<?> that)
             throws ConversionException {
+
+        // Cache the converters for ProductUnit because they are expensive to calculate.
+        // For the other units this computation is very fast and adding all of them in the cache reduces the overall
+        // performance (experimentally tested with Scalar benchmark and complex signals)
+        if(this instanceof ProductUnit || that instanceof ProductUnit) {
+            ConcurrentHashMap<Unit<?>, UnitConverter> unitConverters = CONVERTER_CACHE.get(this);
+            if (unitConverters == null) {
+                if (CONVERTER_CACHE.size() > CONVERTER_CACHE_SIZE_LIMIT) {
+                    CONVERTER_CACHE.clear();
```

Comment:
Why not use Guava's LoadingCache? This will evict least recently used entries automatically, rather than blowing the whole cache. Seeq heavily depends on Guava anyway, so there would be no harm in adding it here.

Reply:
I think the idea was to use the fastest possible Map. I don't see any performance improvements for `LoadingCache` over `ConcurrentHashMap` [here](https://github.com/google/guava/wiki/CachesExplained). I know @mariusseiceanu tested with Caffeine (which by [some claims](https://programmer.ink/think/caffeine-the-king-of-local-cache.html) is the fastest Java cache) and the results were visibly slower.

Reply:
It's not the access performance I'm concerned about, but the `clear` pattern above. `LoadingCache` would evice the least recently used rather than throwing away the whole map. The `clear` pattern has some pretty bad worst case performance, i.e. when you have just over 100 units constantly in use, the cache would always be empty. I find it hard to convince myself that this worst case wouldn't happen in practice.

Reply:
I agree that the `clear` call is problematic.  

I think trying `LoadingCache` with a max of 16, 32, or 64 `Pairs` is worth benchmarking. 
Also, about the benchmarks, I think our real use cases are not too different.  Think about we're multiplying two series of 1,000,000 samples each. We'd be hitting the cache all the time.  We also don't run many more than 100 calc in parallel so thrashing the cache is not so likely (but it could happen).

If `LoadingCache` is not fast, we could find tweak the settings? https://groups.google.com/g/guava-discuss/c/BnoN9q9fDgw

Reply:
also, these benchmarks keep stating that `Caffeine` is faster than `ConcurrentHashMap` so we could go back to testing that with the proper settings:
https://github.com/ben-manes/caffeine/wiki/Benchmarks

Reply:
> these benchmarks keep stating that Caffeine is faster than ConcurrentHashMap

While Caffeine is faster than Guava, it is only faster than `ConcurrentHashMap` in one specific case and is otherwise slower (as it builds on top). A `computeIfAbsent` is more pessimistic in locking during the lookup, so Caffeine prefers to perform a lock-free prescreen and falling back to a compute if absent / expired / weak-soft GC'd. In this code it uses a racy get-compute-put which is subject to a cache stampede (a common cause of outages if an external lookup). Otherwise the penalty is ~33% read throughput compared to `ConcurrentHashMap` which so far has not been a reported problem.

Reply:
> Seeq heavily depends on Guava anyway, so there would be no harm in adding it here.
I've tested with LoadingCache (exactly the same version we have in Seeq) which is used in many places in Seeq -  it's nice because we can limit the size of the cache. But the performance was better with ConcurrentHashmap.
Regarding `clear`, it was the optimal solution I found for keeping the performance. Implementing an LRU eviction adds its own costs.  I think 100 or 200 entries are enough for Seeq usecases. We don't need to keep all the converters in the cache - it will be quickly rebuilt.  And it can be easily changed with a better structure when we find one. I agree that theoretically there might be a worst-case scenario - in that case, the performance will be the same as what happens now, without the cache.

More background info:
I also tried to have a cache of converters in each unit (no indirection cache) but this did not work because Seeq keeps instantiating the "same" Product unit over and over again when multiplying signals. Andres and Eric are working on this. When ready, we can refactor this global cache in cache per unit.

But until then we get improved performance with this PR. I'm thinking of an evolutionary approach.